libfreerdp/core/peer.c

```diff
@@ -1391,81 +1325,87 @@ static BOOL freerdp_peer_send_server_redirection_pdu(
 
 	if (redirFlags & LB_USERNAME)
 	{
-		Stream_Write_UINT32(s, userNameLength);
-		Stream_Write(s, userNameW, userNameLength);
-		free(userNameW);
+		if (!redir_write_string(LB_USERNAME, s, userName))
+			goto fail;
 	}
 
 	if (redirFlags & LB_DOMAIN)
 	{
-		Stream_Write_UINT32(s, domainLength);
-		Stream_Write(s, domainW, domainLength);
-		free(domainW);
+		if (!redir_write_string(LB_DOMAIN, s, domain))
+			goto fail;
 	}
 
 	if (redirFlags & LB_PASSWORD)
 	{
-		Stream_Write_UINT32(s, passwordLength);
-		Stream_Write(s, passwordW, passwordLength);
-		free(passwordW);
+		if (!redir_write_string(LB_PASSWORD, s, password))
+			goto fail;
 	}
 
 	if (redirFlags & LB_TARGET_FQDN)
 	{
-		Stream_Write_UINT32(s, targetFQDNLength);
-		Stream_Write(s, targetFQDNW, targetFQDNLength);
-		free(targetFQDNW);
+		if (!redir_write_string(LB_TARGET_FQDN, s, targetFQDN))
+			goto fail;
 	}
 
 	if (redirFlags & LB_TARGET_NETBIOS_NAME)
 	{
-		Stream_Write_UINT32(s, targetNetBiosNameLength);
-		Stream_Write(s, targetNetBiosNameW, targetNetBiosNameLength);
-		free(targetNetBiosNameW);
+		if (!redir_write_string(LB_TARGET_NETBIOS_NAME, s, targetNetBiosName))
+			goto fail;
 	}
 
 	if (redirFlags & LB_CLIENT_TSV_URL)
 	{
-		Stream_Write_UINT32(s, tsvUrlLength);
-		Stream_Write(s, tsvUrl, tsvUrlLength);
+		if (!redir_write_data(LB_CLIENT_TSV_URL, s, tsvUrlLength, tsvUrl))
+			goto fail;
+	}
+
+	if (redirFlags & LB_REDIRECTION_GUID)
+	{
+		if (!redir_write_data(LB_REDIRECTION_GUID, s, redirectionGuidLength, redirectionGuid))
+			goto fail;
+	}
+
+	if (redirFlags & LB_TARGET_CERTIFICATE)
+	{
+		if (!redir_write_data(LB_REDIRECTION_GUID, s, targetCertificateLength, targetCertificate))
+			goto fail;
 	}
 
 	if (redirFlags & LB_TARGET_NET_ADDRESSES)
 	{
 		UINT32 i;
-		Stream_Write_UINT32(s, targetNetAddressesLength);
+		const size_t start = Stream_GetPosition(s);
+
+		if (!Stream_EnsureRemainingCapacity(s, 2 * sizeof(UINT32)))
+			goto fail;
+
+		Stream_Seek_UINT32(s); /* length of field */
 		Stream_Write_UINT32(s, targetNetAddressesCount);
 		for (i = 0; i < targetNetAddressesCount; i++)
 		{
-			Stream_Write_UINT32(s, targetNetAddressesWLength[i]);
-			Stream_Write(s, targetNetAddressesW[i], targetNetAddressesWLength[i]);
-			free(targetNetAddressesW[i]);
+			if (!redir_write_string(LB_TARGET_NET_ADDRESSES, s, targetNetAddresses[i]))
+				goto fail;
 		}
-		free(targetNetAddressesW);
-		free(targetNetAddressesWLength);
+
+		/* Write length field */
+		const size_t end = Stream_GetPosition(s);
+		Stream_SetPosition(s, start);
+		Stream_Write_UINT32(s, end - start - 8);
```

Comment:
When writing the length of `TartetNetAddresses`, it just shouldn't count `TargetNetAddressesLentght` field which is 4 bytes, not 8, so assuming we are counting bytes I think that should be: `end - start - 4`